firepit/firepit.js

```diff
@@ -0,0 +1,568 @@
+// Copyright 2018, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const fs = require("fs");
+const path = require("path");
+const { fork, spawn } = require("child_process");
+const homePath = require("user-home");
+const chalk = require("chalk");
+const shell = require("shelljs");
+shell.config.silent = true;
+
+const runtime = require("./runtime");
+const version = require("./package.json").version;
+
+let config;
+try {
+  config = require("./config");
+} catch (err) {
+  console.warn("Invalid Firepit configuration, this may be a broken build.");
+  process.exit(2);
+}
+
+function SetWindowTitle(title) {
+  if (process.platform === "win32") {
+    process.title = title;
+  } else {
+    process.stdout.write("\x1b]2;" + title + "\x1b\x5c");
+  }
+}
+
+const isWindows = process.platform === "win32";
+const installPath = path.join(homePath, ".cache", "firebase", "tools");
+let runtimeBinsPath = path.join(homePath, ".cache", "firebase", "runtime");
+
+const npmArgs = [
+  `--script-shell=${runtimeBinsPath}/shell${isWindows ? ".bat" : ""}`,
+  `--globalconfig=${path.join(runtimeBinsPath, "npmrc")}`,
+  `--userconfig=${path.join(runtimeBinsPath, "npmrc")}`,
+  `--scripts-prepend-node-path=auto`
+];
+
+let safeNodePath;
+const unsafeNodePath = process.argv[0];
+
+const flagDefinitions = [
+  "file-debug",
+  "log-debug",
+  "disable-write",
+  "runtime-check",
+  "setup-check",
+  "force-setup",
+  "force-update",
+  "ignore-embedded-cache"
+];
+
+const flags = flagDefinitions.reduce((flags, name) => {
+  flags[name] = process.argv.indexOf(`--tool:${name}`) !== -1;
+  if (flags[name]) {
+    process.argv.splice(process.argv.indexOf(`--tool:${name}`), 1);
+  }
+
+  return flags;
+}, {});
+
+if (flags["runtime-check"]) {
+  console.log(`firepit invoked for runtime check, exiting subpit.`);
+  return;
+}
+
+debug(`Welcome to firepit v${version}!`);
+
+(async () => {
+  const isTopLevel = !process.env.FIREPIT_VERSION;
+  safeNodePath = await getSafeCrossPlatformPath(isWindows, process.argv[0]);
+  uninstallLegacyFirepit();
+
+  if (flags["setup-check"]) {
+    const bins = FindTool("firebase-tools/lib/bin/firebase");
+
+    for (const bin of bins) {
+      bins[bin] = await getSafeCrossPlatformPath(bins[bin]);
+    }
+
+    console.log(JSON.stringify({ bins }));
+    return;
+  }
+
+  if (flags["force-update"]) {
+    console.log(`Please wait while we clear npm's cache...`);
+    process.argv = [
```

Comment:
Ok so it seems like a lot of these "imitation" parts rely on re-assigning `process.argv` and then forking.  Can you explain how that works and why it has to be done by messing with your own `argv`?